### ISSUE TITLE
Switch to crossbeam-channel over std::sync::mpsc?

### ISSUE THREAD

**Issue by tobz** [brayniac/tic#70]
Given that channels are central to everything about tic, I started looking into replacing the `std::sync::mpsc` usage with `crossbeam-channel`, which aims to be (much) faster than `std::sync::mpsc` and others.

I converted over `mio-extras` to use it, and results are promising: I was hitting throughputs of 13.5Msamples/sec (compared to 12.6Msamples/sec normally) and latency p50 dropped from 125ns to _**73ns!**_

For some reason, though, the benchmark throughput takes a nosedive after 20-30 seconds of runtime, dropping to 4-5Msamples/sec, with little bursts back up to 8-9Msamples/sec, but ultimately never recovering.  I haven't been able to work that out yet.

Just figured I'd mention it here in case you had thoughts or had already attempted this.  I'd post a PR and comment on that instead but it would be a PR for `mio-extras` so we'd have no good way to collaborate on it, I don't think.

**Comment by brayniac** [comment]
Interesting results! The drop in throughput is a little concerning. I am a bit hesitant to swap out the channel implementation at this stage unless we are able to obtain consistently solid performance.

**Comment by tobz** [comment]
Long time, no pester! :)

I think I effectively pinned this down to the memory usage of `tic` in general.  I haven't fully tested it, but, my hypothesis is that because the buffer queue used by senders is unbounded, and senders will directly allocate a new buffer if none are available, once you hit a certain amount of memory pressure, the system gets angry which in turn affects throughput.  This actually happens rather quickly during the benchmark run... I was noticing 8GB memory usage, or 50% of my system, when the test started to slow down.  I think speeding things up with crossbeam-channel actually made it allocate faster, thus hitting the bug, haha!

Weirdly, I've actually forked `tic` and made this crossbeam_channel switch, along other things, and I maintain a fixed set of buffers -- no runtime allocations -- and I haven't noticed the same type of slowdowns that I originally did when testing this change in `tic`.

Given that you said you're working on a rewrite of `tic` that you eventually plan to release, I'm not sure how much sense it makes to try and push forward on this, given that it would require extracting the channel portion of `mio-extras` (since I don't think they would take such a patch) and all that... but I figured I would at least let you know and close the loop, in case it happened to help you out. :)
